### PR TITLE
Map clouds by provider type

### PR DESCRIPTION
**Pull request by huwshimi**
Fixes: #2095.

### CODE REVIEW

**Comment by huwshimi** [comment]
QA:
- deploy the gui on a bunch of clouds
- check that you can choose clouds and add credentials and the logos etc. appear

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5942/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5943/
Test FAILed.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -1097,7 +1097,7 @@ YUI.add('juju-gui', function(Y) {
       // Auto place the units. This is probably not the best UX, but is required
       // to display the machines in the deployment flow.
       this._autoPlaceUnits();
-      let cloud = env.get('cloud');
+      let cloud = env.get('providerType');
       if (cloud) {
         cloud = {name: cloud};
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: this is no longer the `name` it should probably be `type`

File: jujugui/static/gui/src/app/components/deployment-flow/cloud/cloud.js [hunk 2]
```diff
@@ -125,7 +125,7 @@ YUI.add('deployment-cloud', function() {
       @returns {Array} The logo.
     */
     _generateLogo: function(cloud) {
-      var info = this.props.clouds[cloud.name];
+      var info = this.props.providers[cloud.cloudType];
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-flow/cloud/cloud.js]: where does cloudType get into the cloud object?

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/deployment-flow/cloud/cloud.js]: https://github.com/juju/juju-gui/blob/develop/jujugui/static/gui/src/app/store/env/controller-api.js#L726

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/app.js]: Oops

Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: If this worked...maybe it's not being used?

Comment (huwshimi) [inline jujugui/static/gui/src/app/app.js]: There was a bug that allowed it to work that I'm currently fixing.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5944/
Test PASSed.

File: jujugui/static/gui/src/app/app.js [hunk 3]
```diff
@@ -1097,9 +1097,12 @@ YUI.add('juju-gui', function(Y) {
       // Auto place the units. This is probably not the best UX, but is required
       // to display the machines in the deployment flow.
       this._autoPlaceUnits();
-      let cloud = env.get('cloud');
+      let cloud = env.get('providerType');
```
Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: s/let cloud/let cloudType/

File: jujugui/static/gui/src/app/components/deployment-flow/cloud/cloud.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-flow/cloud/cloud.js]: const, while you are there?

File: jujugui/static/gui/src/app/components/deployment-flow/credential/add/add.js [hunk 4]
```diff
@@ -25,16 +25,16 @@ YUI.add('deployment-credential-add', function() {
...
       getCredentials: React.PropTypes.func.isRequired,
+      providers: React.PropTypes.object.isRequired,
       setCredential: React.PropTypes.func.isRequired,
       updateCloudCredential: React.PropTypes.func.isRequired,
       user: React.PropTypes.string,
       validateForm: React.PropTypes.func.isRequired
     },
 
-    DEFAULTCLOUD: 'google',
+    DEFAULTCLOUD: 'gce',
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-flow/credential/add/add.js]: DEFAULT_CLOUD_TYPE ?

**Comment by urosj** [comment]
QA on openstack is 👍 for this branch, but 👎overall. Domain name is not required, please leave it optional.

**Comment by hatched** [comment]
Tested in AWS and GCE and alls well.

Here is the issue for the domain not being required. https://github.com/juju/juju-gui/issues/2102 Because this is a separate issue I'd be ok with a follow-up branch.

File: jujugui/static/gui/src/app/app.js (hunk 3 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/app.js]: The env only has `providerType`, I couldn't see anything about a `cloudType`, not sure if that's a bug.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5953/
Test PASSed.

**Comment by huwshimi** [comment]
Thanks everyone :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**